### ISSUE TITLE
Investigate constrained JSON output (Anthropic/OpenAI) to enforce schema compliance at token level

### ISSUE THREAD

**Issue by wphillipmoore** [wphillipmoore/ai-research-methodology#107]
## Problem

LLMs do not reliably respect JSON schemas provided in prompts. The model treats the schema as guidance, not as a constraint — it adds "helpful" fields not in the schema, returns null for string-typed fields, and invents metadata. Each violation requires either a schema patch (reactive, infinite loop) or a Python-side workaround.

On feature/101, we implemented pre-validation field stripping (`_strip_to_schema` in `api_client.py`) as the strategic Python-side fix. This works but it would be better to not need it at all.

## Investigation

OpenAI offers `response_format: { type: "json_schema", json_schema: {...} }` which enforces schema compliance at the **token-generation level** — the model literally cannot produce non-schema tokens (constrained decoding). Does Anthropic offer an equivalent?

Areas to check:
- Anthropic API docs for structured output / JSON mode
- `tool_use` response format constraints (we already use tool_use for some calls)
- Third-party wrappers (instructor, outlines, guidance) that add constrained decoding to Anthropic models
- Anthropic roadmap signals on structured outputs
- Whether this is a practical reason to evaluate multi-vendor support (using OpenAI's structured output for schema-sensitive steps while keeping Anthropic for judgment-heavy steps)

## Why this matters

Schema violations are not rare edge cases. In a typical single-query research run (15 sources, 10+ sub-agent calls), we observe 2-5 schema violations per run. Each one historically required a re-run ($2+) to discover and a schema patch to work around. The stripping fix eliminates the re-run cost but the violations still represent the LLM ignoring explicit instructions — which undermines trust in all other instructions it was given.

If constrained decoding exists and works, it eliminates the entire class of problems and makes `_strip_to_schema` unnecessary.

## Priority

Medium. The stripping fix is working. This is an optimization that would eliminate the need for it.

## Related

- feature/101 (pipeline events) — where the stripping was implemented
- `project_show_your_work_principle` — deterministic enforcement over probabilistic compliance
- the-infrastructure-mindset #153 (memory WTF) — same theme: AI infrastructure that doesn't enforce its own contracts

**Comment by wphillipmoore** [comment]
**Research findings (2026-04-20): Anthropic HAS structured outputs — GA, token-level enforcement.**

## The answer

Anthropic's Claude API supports **constrained decoding** for JSON schema enforcement. It's generally available (no beta header needed) across Claude Opus 4.7, Opus 4.6, Sonnet 4.6, Sonnet 4.5, Haiku 4.5. The schema is compiled into a context-free grammar (CFG) that constrains which tokens can be generated — invalid tokens have their logits set to negative infinity. Schema violations are **mathematically impossible**.

## Two paths, both token-level enforcement

**Path 1 — Direct JSON output (`output_config.format`):**
```python
response = client.messages.create(
    model="claude-sonnet-4-20250514",
    max_tokens=8192,
    messages=[...],
    output_config={
        "format": {
            "type": "json_schema",
            "schema": {
                "type": "object",
                "properties": {...},
                "required": [...],
                "additionalProperties": False
            }
        }
    }
)
```

**Path 2 — Strict tool use (`strict: true`):**
```python
tools=[{
    "name": "source_scorer",
    "strict": True,
    "input_schema": { ... your schema ... }
}]
```

## Performance

- Grammar compilation adds latency on the **first request** per schema
- Compiled grammars are **cached 24 hours** from last use
- Subsequent requests with the same schema are fast
- Since we reuse the same schemas across all calls in a run, only the first call per schema type pays the compilation cost

## What this eliminates

Once implemented:
- The model **cannot** produce `source_title` or any non-schema field (the outlier issue from run 4)
- The model **cannot** produce `"date": null` when the schema says `type: "string"` (the nullable issue)
- The `_strip_to_schema` safety net becomes provably unnecessary (tracked separately)
- Every sub-agent call returns guaranteed-valid JSON matching its schema — no retries, no patching, no stripping

## Implementation plan

Update `api_client.py`'s `call_sub_agent()` to pass `output_config` with the schema when calling the API. Our schemas are already loaded for validation — we just need to pass them to the API as well. Minimal code change:

```python
# Before (current): schema used for post-hoc validation only
response = self._client.messages.create(
    model=model,
    max_tokens=max_tokens,
    system=system_prompt,
    messages=[{"role": "user", "content": user_content}],
)

# After: schema enforced at token generation AND post-hoc validated
response = self._client.messages.create(
    model=model,
    max_tokens=max_tokens,
    system=system_prompt,
    messages=[{"role": "user", "content": user_content}],
    output_config={
        "format": {
            "type": "json_schema",
            "schema": schema_dict,
        }
    },
)
```

## Sources

- [Structured outputs - Claude API Docs](https://platform.claude.com/docs/en/build-with-claude/structured-outputs)
- [Strict tool use - Claude API Docs](https://platform.claude.com/docs/en/agents-and-tools/tool-use/strict-tool-use)
- [Anthropic Claude Tutorial: Structured Outputs with Instructor](https://python.useinstructor.com/integrations/anthropic/)
